solvation_analysis/solution.py

```diff
@@ -0,0 +1,370 @@
+        """
+        Returns the solvation shell of the solute as an AtomGroup.
+
+        Parameters
+        ----------
+        solute_index : int
+            The index of the solute of interest
+        frame : int
+            the frame in the trajectory to perform selection at. Defaults to the
+            current trajectory frame.
+        as_df : bool, default False
+            if true, this function will return a DataFrame representing the shell
+            instead of a AtomGroup.
+        remove_mols : dict of {str: int}, optional
+            remove_dict lets you remove specific residues from the final shell.
+            It should be a dict of molnames and ints e.g. {'mol1': n, 'mol2', m}.
+            It will remove up to n of mol1 and up to m of mol2. So if the dict is
+            {'mol1': 1, 'mol2', 1} and the shell has 4 mol1 and 0 mol2,
+            solvation_shell will return a shell with 3 mol1 and 0 mol2.
+        closest_n_only : int, optional
+            if given, only the closest n residues will be included
+
+        Returns
+        -------
+        MDAnalysis.AtomGroup or DataFrame
+
+        """
+        assert self.solvation_frames, "Solute.run() must be called first."
+        assert frame in self.frames, ("The requested frame must be one "
+                                      "of an analyzed frames in self.frames.")
+        remove_mols = {} if remove_mols is None else remove_mols
+        # select shell of interest
+        shell = self.solvation_data.xs((frame, solute_index), level=("frame", "solvated_atom"))
+        # remove mols
+        for mol_name, n_remove in remove_mols.items():
+            # first, filter for only mols of type mol_name
+            is_mol = shell.res_name == mol_name
+            res_ids = shell[is_mol].res_id
+            mol_count = len(res_ids)
+            n_remove = min(mol_count, n_remove)
+            # then truncate resnames to remove mols
+            remove_ids = res_ids[(mol_count - n_remove):]
+            # then apply to original shell
+            remove = shell.res_id.isin(remove_ids)
+            shell = shell[np.invert(remove)]
+        # filter based on length
+        if closest_n_only:
+            assert closest_n_only > 0, "closest_n_only must be at least 1"
+            closest_n_only = min(len(shell), closest_n_only)
+            shell = shell[0: closest_n_only]
+        if as_df:
+            return shell
+        else:
+            return self._resids_to_atom_group(shell["res_id"], solute_index=solute_index)
+
+    def _resids_to_atom_group(self, ids, solute_index=None):
+        """
+
+        Parameters
+        ----------
+        ids : numpy.array of int
+            an array of res ids
+        solute_index : int, optional
+            if given, will include the solute with solute_index
+
+        Returns
+        -------
+        MDAnalysis.AtomGroup
+        """
+        ids = " ".join(ids.astype(str))
```

Comment:
You should construct directly from an array or slice rather than using the selection API.
Something along the lines of 

```
indices = [0,1,2,3,4,5]
ag = self.u.atoms[indicies]
```

Reply:
Ahh yes this is much cleaner.